test/e2e/framework/service/jig.go

```diff
@@ -390,8 +393,52 @@ func (j *TestJig) waitForAvailableEndpoint(timeout time.Duration) error {
 
 	go controller.Run(stopCh)
 
+	// If EndpointSlice API is enabled, then validate if appropriate EndpointSlice objects were also create/updated/deleted.
+	if _, err := j.Client.Discovery().ServerResourcesForGroupVersion(discoveryv1beta1.SchemeGroupVersion.String()); err == nil {
+		var esController cache.Controller
+		_, esController = cache.NewInformer(
+			&cache.ListWatch{
+				ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
+					opts := metav1.ListOptions{
+						LabelSelector: "kubernetes.io/service-name=" + j.Name,
+					}
+					obj, err := j.Client.DiscoveryV1beta1().EndpointSlices(j.Namespace).List(context.TODO(), opts)
+					return runtime.Object(obj), err
+				},
+				WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+					opts := metav1.ListOptions{
+						LabelSelector: "kubernetes.io/service-name=" + j.Name,
+					}
+					return j.Client.DiscoveryV1beta1().EndpointSlices(j.Namespace).Watch(context.TODO(), opts)
+				},
+			},
+			&discoveryv1beta1.EndpointSlice{},
+			0,
+			cache.ResourceEventHandlerFuncs{
+				AddFunc: func(obj interface{}) {
+					if es, ok := obj.(*discoveryv1beta1.EndpointSlice); ok {
+						if len(es.Endpoints) > 0 && len(es.Endpoints[0].Addresses) > 0 {
```

Comment:
Is this really the logic we would like to have? There can be more than a single EndpointSlice object.

Reply:
Actually - looking into Endpoints logic above it seems we also don't wait until Endpoints is fully propagated, but for it to exist.

That's not ideal imho, but it's at least consistent.

Reply:
nah, we should fix it.
is enough to check that the endpoints are ready?

Reply:
Technically, i can imagine a service with 1000 pods, but endpoint/endpointslice at this point containing a single one.
I don't think we have enough information at this point to add a comprehensive check.

I would start with something (e.g. what you have) and drop a TODO to introduce a more comprehensive check and wouldn't block this PR on that.

Reply:
done